src/main/scala/org/broadinstitute/dsde/workbench/gpalloc/dao/HttpGoogleBillingDAO.scala

```diff
@@ -0,0 +1,337 @@
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.api.services.admin.directory.DirectoryScopes
+import com.google.api.services.cloudbilling.Cloudbilling
+import com.google.api.services.cloudbilling.model.ProjectBillingInfo
+import com.google.api.services.cloudresourcemanager.CloudResourceManager
+import com.google.api.services.cloudresourcemanager.model.{Policy, Project, SetIamPolicyRequest}
+import com.google.api.services.compute.model.UsageExportLocation
+import com.google.api.services.compute.{Compute, ComputeScopes}
+import com.google.api.services.genomics.GenomicsScopes
+import com.google.api.services.iam.v1.Iam
+import com.google.api.services.iam.v1.model.{ServiceAccount, ServiceAccountKey}
+import com.google.api.services.plus.PlusScopes
+import com.google.api.services.servicemanagement.ServiceManagement
+import com.google.api.services.servicemanagement.model.EnableServiceRequest
+import com.google.api.services.storage.model.Bucket.Lifecycle
+import com.google.api.services.storage.model.Bucket.Lifecycle.Rule.{Action, Condition}
+import com.google.api.services.storage.model.{Bucket, BucketAccessControl, ObjectAccessControl}
+import com.google.api.services.storage.{Storage, StorageScopes}
+import io.grpc.Status.Code
+import org.broadinstitute.dsde.workbench.google.GoogleUtilities
+import org.broadinstitute.dsde.workbench.gpalloc.db.ActiveOperationRecord
+import org.broadinstitute.dsde.workbench.gpalloc.model.BillingProjectStatus._
+import org.broadinstitute.dsde.workbench.gpalloc.model.{BillingProjectStatus, GPAllocException}
+import org.broadinstitute.dsde.workbench.metrics.{GoogleInstrumented, GoogleInstrumentedService}
+import org.broadinstitute.dsde.workbench.model.{ErrorReport, ErrorReportSource, WorkbenchExceptionWithErrorReport}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+
+class HttpGoogleBillingDAO(appName: String, clientSecrets: GoogleClientSecrets, serviceAccountPemFile: String)
+                           (implicit val system: ActorSystem, val executionContext: ExecutionContext)
+  extends GoogleUtilities {
+
+  protected val workbenchMetricBaseName = "billing"
+  implicit val service = GoogleInstrumentedService.Iam
+
+  implicit val counters = googleCounters(service)
+  implicit val histo = googleRetryHistogram(service)
+
+  implicit val errorReportSource = ErrorReportSource("gpalloc-google")
+
+  val serviceAccountClientId: String = clientSecrets.getDetails.get("client_email").toString
+
+  lazy val httpTransport = GoogleNetHttpTransport.newTrustedTransport
+  lazy val jsonFactory = JacksonFactory.getDefaultInstance
+
+  //giant bundle of scopes we need
+  val saScopes = Seq(
+    StorageScopes.DEVSTORAGE_FULL_CONTROL,
+    ComputeScopes.COMPUTE,
+    DirectoryScopes.ADMIN_DIRECTORY_GROUP,
+    GenomicsScopes.GENOMICS,
+    "https://www.googleapis.com/auth/cloud-billing",
+    PlusScopes.USERINFO_EMAIL,
+    PlusScopes.USERINFO_PROFILE)
```

Comment:
do we really need all these scopes? i would have thought maybe just cloud-billing

Reply:
yeah. each one of the Builders below has a different set of scopes it requires. i just chucked them all into one array so we can build a single mega-credential.